Agentic Commerce: Add support for webhooks spec

ebinnion opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR implements webhook support for the Agentic Commerce Protocol (ACP), enabling WooCommerce stores to send order lifecycle events to AI agents in ACP-compliant format.

View the spec at this URL: https://github.com/agentic-commerce-protocol/agentic-commerce-protocol/blob/main/spec/openapi/openapi.agentic_checkout_webhook.yaml

 Key Features:
  - Webhooks only fire for orders created through Agentic checkout (identified by session meta)
  - Duplicate event prevention for status changes and refunds
  - Full refund history included in update events
  - Extensible status mapping for custom order statuses

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

1. Setup a test webhook:
    - Go to WooCommerce → Settings → Advanced → Webhooks
    - Click "Add webhook"
    - Set Topic to "Agentic Order Created" or "Agentic Order Updated"
    - Set Delivery URL to a test endpoint (e.g., webhook.site)
    - Add a Secret for signature verification (e.g., "my_webhook_secret")
    - Set Status to "Active"
    - Save the webhook

2. Create test products using WP-CLI:
    ```bash
    wp-env run cli wp wc product create --user=1 --name="Gaming Mouse" --type=simple --regular_price=59.99 --porcelain
    wp-env run cli wp wc product create --user=1 --name="Mechanical Keyboard" --type=simple --regular_price=129.99 --porcelain
    ```
    Note the product IDs returned (e.g., 226, 227).

3. Create an Agentic order with WP-CLI:
    ```bash
    wp-env run cli wp wc shop_order create --user=1 --status=pending --line_items='[{"product_id":226,"quantity":2},{"product_id":227,"quantity":1}]' --meta_data='[{"key":"_agentic_checkout_session_id","value":"session_test_456"}]'
    ```
    Replace `226` and `227` with your product IDs. Note the order ID returned (e.g., 229).

4. Verify webhook payload:
    - Check your webhook endpoint received a POST request
    - Verify payload structure matches ACP format:
    ```json
    {
      "type": "order_create",
      "data": {
        "type": "order",
        "checkout_session_id": "session_test_456",
        "permalink_url": "https://example.com/checkout/order-received/229/...",
        "status": "created",
        "refunds": []
      }
    }
    ```
    - Verify HTTP headers include `Merchant-Signature` (HMAC-SHA256 signature)

5. Test order updates:
    ```bash
    wp-env run cli wp wc shop_order update 229 --user=1 --status=processing
    ```
    - Verify "Agentic Order Updated" webhook fires with `status: "confirmed"`

6. Test refunds:
    ```bash
    wp-env run cli wp wc shop_order_refund create 229 --user=1 --amount=49.99 --reason="Customer requested refund" --api_refund=false
    wp-env run cli wp wc shop_order update 229 --user=1 --status=on-hold
    ```
    - Verify webhook includes refund data with correct type and amount in minor units (e.g., 4999 for $49.99)
    - Verify `status: "manual_review"` for on-hold orders

7. Test non-Agentic orders:
    ```bash
    wp-env run cli wp wc shop_order create --user=1 --status=pending
    ```
    - Verify Agentic webhooks do NOT fire for orders without `_agentic_checkout_session_id` meta

8. Run unit tests:
    ```bash
    pnpm run test:php:env -- tests/php/src/Internal/Admin/Agentic/
    ```

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ebinnion committed:
Agentic Commerce: Initial pass at webhooks support

ebinnion committed:
add empty newline to agentic controller

ebinnion committed:
add empty newline to agentic webhook settings test

ebinnion committed:
Remove unnecessary settings to lighten PR

ebinnion committed:
Update @since tag to 10.3.0

ebinnion committed:
Remove unnecessary hook in agentic webhook manager; add docblock to new hooks

ebinnion committed:
Agentic Webhooks:  Fixes an issue where Agentic webhook actions were firing multiple times for the same event, causing duplicate webhooks to be sent.

  Changes:
  - Remove woocommerce_update_order hook to prevent duplicate firing
    (status changes already trigger woocommerce_order_status_changed)
  - Remove unused handle_order_updated() method
  - Add static processed_events tracking to prevent duplicate firing
  - Add reset_processed_events() method for testing

  Testing:
  - Rewrite AgenticWebhookManagerTest to use native WooCommerce webhooks
  - Add test_action_fired_on_multiple_partial_refunds() to verify each
    refund triggers exactly one webhook
  - Add test_webhook_payload_contains_all_refunds() to verify webhook
    payload includes all refund data with correct amounts and types

  All 27 Agentic tests now pass with 69 assertions.

ebinnion committed:
Remove statuses that might exist

ebinnion committed:
Allow filtering order status mapping; validate against spec

ebinnion committed:
Refactor and create helpers to minimize tests

htdat commented on plugins/woocommerce/src/StoreApi/Utilities/OrderController.php:
```diff
@@ -66,6 +66,14 @@ public function create_order_from_cart() {
+			if ( ! empty( $agentic_session_id ) ) {
+				$order->update_meta_data( '_agentic_checkout_session_id', $agentic_session_id );
+			}
+		}
```
We might not need this as this would be handled in the `/complete` endpoint in PR #61271 

https://github.com/woocommerce/woocommerce/blob/b7ca7a30ce3be32f887bd739b8547ee469704d35/plugins/woocommerce/src/StoreApi/Routes/V1/Agentic/CheckoutSessionsComplete.php#L283-L284

htdat commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,327 @@
+	 */
+	private function should_trigger_webhook( $order ) {
+		// Only trigger for orders with an Agentic checkout session ID.
+		$checkout_session_id = $order->get_meta( '_agentic_checkout_session_id' );
```
Nitpick: There will be `OrderMetaKey::AGENTIC_CHECKOUT_SESSION_ID` to use rather than this plain string `_agentic_checkout_session_id`

htdat commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+		// Try to get the order received page URL (customer-facing)
+		$order_received_url = $order->get_checkout_order_received_url();
+
+		if ( ! empty( $order_received_url ) ) {
+			return $order_received_url;
+		}
+
+		// Fallback to admin edit URL if customer URL not available
+		return admin_url( 'post.php?post=' . $order->get_id() . '&action=edit' );
```
I don't see any case that `$order->get_checkout_order_received_url()` fails so we do not really need this fallback.

RadoslavGeorgiev replied:
Technically, someone could use the `woocommerce_get_checkout_order_received_url` filter and return something different from the order URL. But yes, we should not expose the order edit URL. Let's remove this method altogether and directly use `$order->get_checkout_order_received_url()`.

Even if we did, there is the `get_edit_order_url()` method of orders, but that's beside the point.

htdat reviewed (commented):
Some initial comments

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticController.php:
```diff
@@ -0,0 +1,31 @@
+	 *
+	 * This follows the WooCommerce pattern for controllers.
+	 */
+	public function register() {
```
Maybe we can use the opportunity to check whether the feature is enabled, similarly to https://github.com/woocommerce/woocommerce/blob/d1333590556bfda3d9daf72355ee67a284281e5c/plugins/woocommerce/src/StoreApi/RoutesController.php#L94?

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+	 * @param WC_Order $order Order object.
+	 * @return array Webhook payload.
+	 */
+	public function build_payload( $event, $order ) {
```
Not strictly related to this PR, but Woo requires PHP >= 7.4. This means that we can use strong typing for parameters.

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+	 */
+	private function get_checkout_session_id( $order ) {
+		$session_id = $order->get_meta( '_agentic_checkout_session_id' );
+		return ! empty( $session_id ) ? $session_id : 'checkout_session_' . $order->get_id();
```
Is there a scenario where an order might not have the session ID in meta? I'd rather throw an exception and terminate the process here.

FWIW the meta is already required in `AgenticWebhookManager->should_trigger_webhook()` so I don't see a reason to support scenarios with missing `_agentic_checkout_session_id`.

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+	 */
+	private function map_order_status( $wc_status ) {
+		// Remove 'wc-' prefix if present.
+		$wc_status = str_replace( 'wc-', '', $wc_status );
```
Why? I assume that the order status should be percise.

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+
+		$status_map = array(
+			// WooCommerce status => ACP status.
+			'pending'        => ACPOrderStatus::CREATED,
+			'processing'     => ACPOrderStatus::CONFIRMED,
+			'on-hold'        => ACPOrderStatus::MANUAL_REVIEW,
+			'completed'      => ACPOrderStatus::FULFILLED,
+			'cancelled'      => ACPOrderStatus::CANCELED,
+			'canceled'       => ACPOrderStatus::CANCELED, // Support both spellings.
+			'refunded'       => ACPOrderStatus::FULFILLED, // Refunded orders are still fulfilled.
+			'failed'         => ACPOrderStatus::CANCELED,
+			'checkout-draft' => ACPOrderStatus::CREATED,
+		);
```
If we stop removing the `wc-` status above, we can use the stricter constants:
```suggestion
		$status_map = array(
			// WooCommerce status => ACP status.
			OrderStatus::PENDING        => ACPOrderStatus::CREATED,
			OrderStatus::PROCESSING     => ACPOrderStatus::CONFIRMED,
			OrderStatus::ON_HOLD        => ACPOrderStatus::MANUAL_REVIEW,
			OrderStatus::COMPLETED      => ACPOrderStatus::FULFILLED,
			OrderStatus::CANCELLED      => ACPOrderStatus::CANCELED,
			OrderStatus::REFUNDED       => ACPOrderStatus::FULFILLED, // Refunded orders are still fulfilled.
			OrderStatus::FAILED         => ACPOrderStatus::CANCELED,
			OrderStatus::CHECKOUT_DRAFT => ACPOrderStatus::CREATED,
		);
```

> `// Support both spellings.`

Why?

> `// Refunded orders are still fulfilled.`

Yup. Other systems separate the types of statuses, one of the pain points with Woo.

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+		 *
+		 * Allows extensions to map custom WooCommerce order statuses to ACP order statuses.
+		 * The mapped status must be one of: created, manual_review, confirmed, canceled, shipped, fulfilled.
+		 *
```
```suggestion
		 *
         * @see Automattic\WooCommerce\StoreApi\Routes\V1\Agentic\Enums\Specs\OrderStatus
```

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+		if ( ! ACPOrderStatus::is_valid( $mapped_status ) ) {
+			// Log a warning for invalid status but continue with fallback.
+			// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
+			error_log(
```
I think we should not bypass the Woo logger. There should be a `WC_Logger_Interface $logger` property of the class, and we should call `$this->logger->error()`.

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+			$refunds_data[] = $this->build_single_refund_data( $refund );
+		}
+
+		return $refunds_data;
```
Nitpick mode engaged... 😃

I think this can be way shorter:

```suggestion
		return array_map(
			[ $this, 'build_single_refund_data' ],
			$order->get_refunds()
		);
```

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+	 */
+	private function build_single_refund_data( $refund ) {
+		$refund_type = $this->determine_refund_type( $refund );
+		$amount      = abs( (float) $refund->get_total() ); // Get absolute value as refunds are negative
```
We need cents here. The money formatter can be used similarly to [here](https://github.com/woocommerce/woocommerce/blob/d1333590556bfda3d9daf72355ee67a284281e5c/plugins/woocommerce/src/StoreApi/Schemas/V1/Agentic/CheckoutSessionSchema.php#L467-L470).

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+		$refund_type = $refund->get_meta( '_refund_type' );
+
+		if ( 'store_credit' === $refund_type ) {
+			return 'store_credit';
```
```suggestion
			return RefundType::STORE_CREDIT;
```

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+		// Check if refund has a specific type meta
+		$refund_type = $refund->get_meta( '_refund_type' );
+
+		if ( 'store_credit' === $refund_type ) {
```
[This](https://github.com/woocommerce/woocommerce/blob/d1333590556bfda3d9daf72355ee67a284281e5c/plugins/woocommerce/src/Admin/API/Reports/Products/DataStore.php#L120-L150) is the only place where refund type is set, and AFAICT the only options are `full` or `partial`. We'll need to handle things differently here.

ebinnion replied:
Our options for the spec seem to be either `store_credit` or `original_payment`. That said, I've updated the code to default to `original_payment` and added a filter to allow extensions to hook in to change it to `store_credit`.

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+
+		// Check refund reason for store credit indicators
+		$reason = strtolower( $refund->get_reason() );
+		if ( strpos( $reason, 'store credit' ) !== false || strpos( $reason, 'credit' ) !== false ) {
```
Where does this come from? Checking the refund reason seems fragile, especially when you consider that it could be translated. This would not work with "кредит в магазина" (Bulgarian translation).

Also (though I imagine the string is different), what if the reason contains "Credit card"?

ebinnion replied:
It was an attempt by Claude to try to determine if the refund was to store credit. I just removed it and added a filter for now in case extensions want to hook in.

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+	 * @return string Formatted amount.
+	 */
+	private function format_amount( $amount ) {
+		return number_format( $amount, 2, '.', '' );
```
Same as the money formatter comment above.

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,230 @@
+	 * @param WC_Order $order Order object.
+	 * @return array Additional metadata.
+	 */
+	private function get_additional_metadata( $order ) {
```
This is a private method that is not used anywhere. Even if it was, I don't see it to be allowed in the spec.

RadoslavGeorgiev reviewed (changes requested):
This is quite solid for an initial phase of the PR! Kudos, @ebinnion 🙇 

I'm posting my preliminary comments after looking at most of the changes. I've still got to review the `AgenticWebhookManager` class and perform manual tests.

RadoslavGeorgiev commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,327 @@
+		add_filter( 'woocommerce_webhook_topics', array( $this, 'register_webhook_topic_names' ) );
+
+		// Hook into order lifecycle events to fire our custom actions.
+		add_action( 'woocommerce_new_order', array( $this, 'handle_order_created' ), 10, 2 );
```
I'm wondering whether it makes much sense to use a hook to trigger the order creation webhook.

With agentic commerce, we control the flow. This means that we can control exactly when to send the webhook, ex. right before or after the payment while responding to the `checkout_sessions/{id}/complete` endpoint. At the same time, I don't know when the order will be created. Is if before/after checking for errors? Does it matter if all validation passed or not?

My (and Claude's :lol:) understanding is that an `order_created` webhook should only be sent pretty much when the `/complete` call is successful.

Also, extensions might already be hooked on this action, and modify the order. In that case, we need to add our hook with the least (latest) priority. The easiest way to achieve this would be to just do it after the call that would create the order.

RadoslavGeorgiev replied:
Marking as resolved in favor of WOOPTP-23

ebinnion committed:
First pass at addressing lint issues

ebinnion committed:
Further fix lint issues

ebinnion committed:
Add merchant-signature header

ebinnion committed:
Simplify refund type for webhook so we do not guess as much

ebinnion committed:
Remove unused helper

ebinnion committed:
Only load agenticcontroller when agentic_checkout feature is enabled

ebinnion committed:
Add strong typing

ebinnion committed:
Remove status filtering

ebinnion committed:
Remove extra spelling

ebinnion committed:
Use OrderStatus enum for mapping statuses

ebinnion committed:
Simplify building refunds data

ebinnion committed:
Use MoneyFormatter from StoreAPI

ebinnion committed:
Use RefundType enum in payload builder

ebinnion committed:
Use WooCommerce logger

ebinnion committed:
Add @see to status map filter

ebinnion committed:
Remove unnecessary fallback for checkout session id

ebinnion committed:
Add changelog

ebinnion committed:
Use webhook signature method instead of rolling our own

ebinnion committed:
Address lint issues

ebinnion commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticController.php:
```diff
@@ -0,0 +1,38 @@
+ *
+ * @since 10.3.0
+ */
+class AgenticController {
```
Noting that this will need some investigation. Some notes.

--

Use DI + registerable patterns consistently

  - Problem: AgenticController does not implement RegisterHooksInterface and new’s the manager manually.
  - Why it matters: Most internal controllers implement RegisterHooksInterface and are resolved via the container; dependencies flow via container and init methods where needed.
  - Fix:
      - Make AgenticController implement Automattic\WooCommerce\Internal\RegisterHooksInterface.
      - In register(), resolve the manager via the container: $this->webhook_manager = wc_get_container()->get( AgenticWebhookManager::class ); Avoid new to keep creation consistent and testable.
      - If you need dependencies for the manager (e.g., logger), pass them via an init() method to follow DI patterns used elsewhere.

ebinnion replied:
Further notes:

--

 Controller patterns

  - Problem: AgenticController::register() only conditionally instantiates the manager; it doesn’t register hooks itself, nor does it guard against installs/upgrades.
  - Why it matters: Many controllers:
      - Implement RegisterHooksInterface
      - Early-return on WC_INSTALLING
      - Do only hook registration in register(), not instantiation.
  - Fix:
      - Add WC_INSTALLING guard via Automattic\Jetpack\Constants.
      - Keep register() focused on hooking. Dependency instantiation should be left to the container.

ebinnion commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,330 @@
+	 */
+	private function is_delivering_webhook() {
+		// phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_debug_backtrace -- Needed to detect delivery context.
+		$backtrace = debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS, 10 );
```
This should disappear 😄

ebinnion commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,330 @@
+	 */
+	public function register_webhook_events( $events ) {
+		// Add our custom events to the valid events list.
+		$events[] = 'agentic_created';
```
Codex didn't like this. 😄 

Requires a bit more understanding on my part.

--

Don’t register custom events for action topics

  - Problem: register_webhook_events() adds agentic_created/agentic_updated to woocommerce_valid_webhook_events.
  - Why it matters: Custom action topics ‘action.woocommerce_*’ are validated by prefix and don’t use the resource/event validation path. This code is unused.
  - Fix: Remove register_webhook_events(). Keeping it adds confusion with no effect.

ebinnion commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,330 @@
+	 *
+	 * @var array
+	 */
+	private static $processed_events = array();
```
Codex seems to think we don't need this. Needs a bit more investigation.

--

Rely on core webhook duplicate prevention

  - Problem: AgenticWebhookManager tracks processed events statically to prevent duplicates.
  - Why it matters: WC_Webhook already dedupes per webhook in-process by $arg. Your custom action fires twice in a single request only if you explicitly do it; core dedupe will handle repeated triggers for
  the same arg.
  - Fix:
      - Remove static $processed_events and the related logic unless you have a case core dedupe doesn’t cover. Current created/updated/refunded flows shouldn’t need this.

ebinnion commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,192 @@
+
+		// Convert amount to cents using MoneyFormatter, same as checkout session endpoints.
+		$formatter       = new MoneyFormatter();
+		$amount_in_cents = (int) $formatter->format( $amount, array( 'decimals' => 2 ) );
```
Needs further investigation. Notes from Codex.

--

 - Problem: Payload builder forces 2 decimals via MoneyFormatter (decimals => 2).
  - Why it matters: Some currencies are zero or three decimals; WooCommerce already exposes wc_get_price_decimals() for store minor units. If the Agentic spec requires minor units for the store currency, we
  should respect store configuration.
  - Fix:
      - Use wc_get_price_decimals() instead of hardcoding 2.
      - Instantiate MoneyFormatter once (property) rather than per-refund.

ebinnion commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,330 @@
+	 */
+	public function register_webhook_topic_names( $topics ) {
+		$topics['action.woocommerce_agentic_order_created'] = __( 'Agentic Order Created', 'woocommerce' );
+		$topics['action.woocommerce_agentic_order_updated'] = __( 'Agentic Order Updated', 'woocommerce' );
```
Should use constants here.

ebinnion commented on plugins/woocommerce/src/StoreApi/Routes/V1/Agentic/Enums/Specs/OrderStatus.php:
```diff
@@ -0,0 +1,66 @@
+<?php
+declare(strict_types=1);
+namespace Automattic\WooCommerce\StoreApi\Routes\V1\Agentic\Enums\Specs;
```
Should consider moving this. Notes from codex.

--

Placement and naming

  - Problem: “Specs” enums are under StoreApi/Routes/V1/Agentic/Enums/Specs, but they’re used outside Store API (webhooks). The “V1” in the namespace hints they’re API-versioned which they aren’t.
  - Why it matters: Cross-cutting enums tied to the Agentic spec are better placed in a neutral, internal location to avoid accidental coupling to Store API route versioning.
  - Fix:
      - Move specs enums to Internal/Agentic/Enums/Specs or Internal/Admin/Agentic/Enums/Specs. Update imports in the payload builder.

ebinnion committed:
Address more lint issues

ebinnion committed:
Refactor AgenticController to use DI and RegisterHooksInterface

- Implement RegisterHooksInterface for consistent controller pattern
- Add WC_INSTALLING guard using Automattic\Jetpack\Constants
- Use wc_get_container()->get() instead of new for AgenticWebhookManager
- Update test to resolve controller from container
- Handle WC_INSTALLING constant in test environment

This aligns with WooCommerce's established patterns for controllers
and makes the code more testable through proper dependency injection.

ebinnion committed:
Remove debug_backtrace from webhook delivery URL logic

- Remove customize_webhook_delivery_url() method and filter
- Remove is_delivering_webhook() helper that used debug_backtrace
- Integrators should now enter the full ACP endpoint URL directly

This simplifies the code by removing brittle stack inspection and
provides clearer UX - users enter the complete webhook URL rather
than having it programmatically rewritten.

ebinnion committed:
Remove unused register_webhook_events() method

- Remove woocommerce_valid_webhook_events filter hook
- Remove register_webhook_events() method

Custom action topics (action.woocommerce_*) are validated by prefix
and don't use the resource/event validation path, making this code
unused and potentially confusing.

ebinnion committed:
Remove redundant static duplicate prevention logic

- Remove $processed_events static property
- Remove duplicate checks from handle_order_status_changed()
- Remove duplicate checks from handle_order_refunded()
- Remove reset_processed_events() method
- Update tests to remove reset_processed_events() call

WC_Webhook already dedupes per webhook in-process by $arg, making
this static tracking unnecessary. The custom actions will only fire
once per unique order_id in a single request.

ebinnion committed:
Fix money formatting to respect currency minor units

- Add MoneyFormatter property, instantiate once in constructor
- Remove hardcoded 'decimals' => 2 argument
- Let MoneyFormatter use wc_get_price_decimals() by default
- Rename variable to amount_in_minor_units for clarity

This ensures money amounts respect the store's configured currency
decimals (e.g., 0 for JPY, 2 for USD, 3 for BHD) and improves
efficiency by creating formatter once instead of per-refund.

ebinnion committed:
Use constants consistently in register_webhook_topic_names()

- Replace hardcoded topic strings with class constants
- Use self::TOPIC_ORDER_CREATED and self::TOPIC_ORDER_UPDATED

This prevents drift and typos by using the single source of truth
for topic strings defined as class constants.

ebinnion committed:
Move Agentic spec enums from StoreApi to Internal namespace

- Relocate all Specs enums from StoreApi/Routes/V1/Agentic/Enums/Specs
  to Internal/Agentic/Enums/Specs
- Update namespace declarations in all 12 enum files
- Update import statements across codebase (5 files affected)

These enums define the Agentic Commerce Protocol spec and are used
across both webhooks and Store API. Moving them to a neutral Internal
location avoids coupling to Store API route versioning and better
reflects their cross-cutting nature.

ebinnion committed:
Add type hints and improve logging level

- Add array return type to register_webhook_topic_names()
- Change invalid ACP status log from error to warning

The return type improves static analysis. The warning level is more
appropriate since the invalid status is recoverable - we gracefully
fall back to 'created' as a default value.

ebinnion committed:
Hook late on new order

ebinnion committed:
Address lint issue of $refund_id never used

ebinnion committed:
Update @see for enum

htdat commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticController.php:
```diff
@@ -0,0 +1,45 @@
+		}
+
+		// Resolve webhook manager from container.
+		$this->webhook_manager = wc_get_container()->get( AgenticWebhookManager::class );
```
~`AgenticController` makes this feels like a REST API endpoint while it's not really~ (_Edit_: this can be very opinionated from me and does not reflect the current pattern of using suffix `Controller` for non-REST stuff). 

Instead, it implements `RegisterHooksInterface` is just to let internal classes properly register hooks.

At the same time, the process of resolving webhook manager `webhook_manager` eventually is to also register hooks `AgenticWebhookManager::init_hooks` 

I think we can simply this a lot by: 

1. Remove `AgenticController` and keep only `AgenticWebhookManager`
2. Let `AgenticWebhookManager` implements `RegisterHooksInterface`. `AgenticWebhookManager::init_hooks` can be changed to `AgenticWebhookManager::register`.
3. Definitely, we should still keep two checks above regarding `WC_INSTALLING` and feature check `agentic_checkout`.

RadoslavGeorgiev replied:
Inspired by this comment, but slightly different solution in https://github.com/woocommerce/woocommerce/pull/61325/commits/c08df06b579a7850287501fa6c6c53dafaedf300.

1. `AgenticController` remains, following the example of other features in WooCommerce.
2. `AgenticWebhooksManager` now also implements `RegisterHooksInterface`, but `->register()` is only called from the `AgenticController` whenever the feature is available and WC is not installing.
3. I replaced the `new` instantiations with actual calls to the dependency container.

As a bonus, tests are a bit clearer now.

RadoslavGeorgiev replied:
@ebinnion I'm shamelessly touching your code, but I won't be marking threads as resolved. I'll leave this up to you, if you agree with the solution.

RadoslavGeorgiev replied:
Taking it back, I did not see any objections, so I'm marking this one as resolved.

htdat commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,233 @@
+	 */
+	public function __construct() {
+		$logger                = wc_get_logger();
+		$this->payload_builder = new AgenticWebhookPayloadBuilder( $logger );
```
Logger is not being used here so I don't think we need to initialize it here. Let `AgenticWebhookPayloadBuilder` to do that itself should be fine.

RadoslavGeorgiev replied:
In https://github.com/woocommerce/woocommerce/commit/c08df06b579a7850287501fa6c6c53dafaedf300 I switched to calling `wc_get_logger()`, following the example of other classes. The discussion is moot now, marking this one as resolved.

htdat commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,199 @@
+	public function __construct( WC_Logger_Interface $logger ) {
+		$this->logger          = $logger;
+		$this->money_formatter = new MoneyFormatter();
+	}
```
Since this will be resolved via `Containner::get`, this can work. We can also `init` like many other existing classes such as [this one](https://github.com/woocommerce/woocommerce/blob/2645ca8e793d4365c8cd2b85c187f9d5f187dd50/plugins/woocommerce/src/Internal/TransientFiles/TransientFilesEngine.php#L59-L68).

```suggestion
	public function __construct( WC_Logger_Interface $logger, MoneyFormatter $money_formatter ) {
		$this->logger          = $logger;
		$this->money_formatter = $money_formatter;
```

htdat replied:
Ignored this one as it's a bit tricky with the logger.

RadoslavGeorgiev replied:
Again, with the direct call to `wc_get_logger()`, this is not an issue anymore in https://github.com/woocommerce/woocommerce/commit/c08df06b579a7850287501fa6c6c53dafaedf300. The `init()` method still instantiates the formatter, but only because it is so simple.

htdat replied:
Nice one! :D

htdat commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,233 @@
+		add_filter( 'woocommerce_webhook_topics', array( $this, 'register_webhook_topic_names' ) );
+
+		// Hook into order lifecycle events to fire our custom actions.
+		add_action( 'woocommerce_new_order', array( $this, 'handle_order_created' ), 10, 999 ); // Hook late to give a chance for other plugins to modify.
```
```suggestion
		add_action( 'woocommerce_new_order', array( $this, 'handle_order_created' ), 999, 2 ); // Hook late to give a chance for other plugins to modify.
```

RadoslavGeorgiev replied:
I still insist that `woocommerce_new_order` is not too reliable. I'll try to call `handle_order_created` when the order gets created in the Agentic API. Commit incoming soon...

RadoslavGeorgiev replied:
Argh, ok... 😃

I implemented your suggestion in https://github.com/woocommerce/woocommerce/pull/61325/commits/37a9f7dc14889777ac965a9e28868c8ef4da952d.

As far as it comes to calling the create order webhook, I think it makes sense to do after the `$this->order_controller->create_order_from_cart();` call. However, right now that is done in the route, and routes are constructed [here](https://github.com/woocommerce/woocommerce/blob/37a9f7dc14889777ac965a9e28868c8ef4da952d/plugins/woocommerce/src/StoreApi/RoutesController.php#L117-L120) without the opportunity for clean dependency injection. I'll open an issue for this, it doesn't make sense to block the PR for it.

ebinnion replied:
haha. What a failure on my part. Nice catch.

htdat commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,199 @@
+	private function get_order_permalink( WC_Order $order ): string {
+		// Get the order received page URL (customer-facing).
+		return $order->get_checkout_order_received_url();
+	}
```
This seems redundant as it is being used only one time, we can directly use `$order->get_checkout_order_received_url()` in `build_order_data`.

RadoslavGeorgiev replied:
Done in https://github.com/woocommerce/woocommerce/pull/61325/commits/a50676d17ece4ee700827d2e54281ab875579c6b

htdat commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilder.php:
```diff
@@ -0,0 +1,199 @@
+			OrderStatus::CANCELLED      => ACPOrderStatus::CANCELED,
+			OrderStatus::REFUNDED       => ACPOrderStatus::FULFILLED, // Refunded orders are still fulfilled.
+			OrderStatus::FAILED         => ACPOrderStatus::CANCELED,
+			OrderStatus::CHECKOUT_DRAFT => ACPOrderStatus::CREATED,
```
I don't think we should track `CHECKOUT_DRAFT`. Using `PENDING` is quite enough. Right before getting int the `process_payment`, all checkout paths will change order status to `PENDING`.

RadoslavGeorgiev replied:
Done in https://github.com/woocommerce/woocommerce/pull/61325/commits/ebe3213a38eec48a33afb8f78bbcd052bfdbbfd2

htdat commented on plugins/woocommerce/src/Internal/Agentic/Enums/Specs/OrderStatus.php:
```diff
@@ -0,0 +1,66 @@
+	 */
+	const CANCELED = 'canceled';
+
+	/**
+	 * Order has been shipped.
+	 */
+	const SHIPPED = 'shipped';
+
+	/**
+	 * Order has been fulfilled.
+	 */
+	const FULFILLED = 'fulfilled';
```
No issue with this PR. Just my observation, I think these two “shipped” or “fulfilled” are the same, or maybe they do matter with `FulfillmentType`: digital and shipping.

RadoslavGeorgiev replied:
Claude's opinion, confirming what I was thinking:

> While the spec doesn't provide explicit definitions, based on standard e-commerce practices:
> 
> - shipped: The order has been handed off to a carrier and is in transit to the customer. The merchant has completed their fulfillment process and the package is on its way.
> - fulfilled: The order has been completely satisfied - typically meaning the customer has received the items. This is the final successful state of an order.
> 
> Typical Order Flow
> 
> 1. created → Order created from completed checkout session
> 2. confirmed → Order confirmed (payment authorized/captured)
> 3. shipped → Package dispatched to carrier, in transit
> 4. fulfilled → Customer received the order, transaction complete

However, this does not mean much for Woo where the order is completed as soon as it ships. In the future, we could integrate with some extension for this.

htdat commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,233 @@
+		}
+
+		// Don't trigger for draft orders.
+		if ( in_array( $order->get_status(), array( OrderStatus::DRAFT, OrderStatus::AUTO_DRAFT ), true ) ) {
```
I think `CHECKOUT_DRAFT` should be in this list too.

RadoslavGeorgiev replied:
Done in https://github.com/woocommerce/woocommerce/pull/61325/commits/ebe3213a38eec48a33afb8f78bbcd052bfdbbfd2

htdat commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,233 @@
+	 * @return array Modified topics.
+	 */
+	public function register_webhook_topic_names( $topics ): array {
+		$topics[ self::TOPIC_ORDER_CREATED ] = __( 'Agentic Order Created', 'woocommerce' );
+		$topics[ self::TOPIC_ORDER_UPDATED ] = __( 'Agentic Order Updated', 'woocommerce' );
```
IMO, if we can combine these topics in a single one such as `Agentic Order`, or `Agentic Order Operations`, or more specific `Agentic Order Create and Update`. With that, merchants just need to configure a webhook to satisfy the requirement.  And then the event `type` will be handled within `AgenticWebhookPayloadBuilder`. 

**That said, this is a nice enhancement to work later**: ideally, merchants do not need to configure this webhook. Instead, we can create a webhook for merchants after they enable this feature with a proper secret. This is inspired [by WooPay](https://github.com/Automattic/woocommerce-payments/blob/3463ea98431595a2e9dbfd2eaddec8b6603d5f39/includes/woopay/class-woopay-order-status-sync.php#L115-L115).

RadoslavGeorgiev replied:
Created an issue to address this: WOOPTP-24

htdat reviewed (commented):
Great work in figuring out the current webhook feature and hook into the proper points!!! 🎉 

I have a few inline comments, none of them are critical. They're all for aligning with the current pattern in the repo, or some ideas to improve further. 

I am going to test this manually and will provide further feedback. 

_Edit_: Manual tests went very well p1760088580226509/1760050608.927719-slack-C09EWAXSYD9

htdat committed:
Merge branch 'trunk' into update/agentic-commerce-protocol-webhooks

# Conflicts:
#	plugins/woocommerce/src/StoreApi/Utilities/AgenticCheckoutUtils.php

htdat committed:
Fix after the merge, mostly due to Specs/Enums namespace change

htdat committed:
lint-fix

RadoslavGeorgiev committed:
Merge branch 'trunk' into update/agentic-commerce-protocol-webhooks

RadoslavGeorgiev committed:
Fix imports

RadoslavGeorgiev committed:
Fix up controller and manager initialization

- Now both the controller and webnhooks manager implement RegisterHooksInterface.
- The controller only calls the manager's `register()` method when the feature is enabled.
- DI: No `new` keywords anymore. DI is used for all classes, except the logger.
- Following the pattern in other classes, the logger is called directly through wc_get_logger()

 Please enter the commit message for your changes. Lines starting

RadoslavGeorgiev committed:
Fix parameter order for adding actions for new orders

RadoslavGeorgiev committed:
Fix the AgenticWebhookPayloadBuilderTest. I did not know it exists 😃

RadoslavGeorgiev committed:
Remove the get_order_permalink method

RadoslavGeorgiev committed:
Ignore CHECKOUT_DRAFT as a status

RadoslavGeorgiev committed:
Remove the logger form the payload builder test

RadoslavGeorgiev commented on plugins/woocommerce/tests/php/src/Internal/Admin/Agentic/AgenticWebhookPayloadBuilderTest.php:
```diff
@@ -0,0 +1,216 @@
+	 */
+	public function setUp(): void {
+		parent::setUp();
+		// Use the WooCommerce logger for tests.
+		$logger                = wc_get_logger();
+		$this->payload_builder = new AgenticWebhookPayloadBuilder();
+		$this->payload_builder->init();
+	}
```
Done https://github.com/woocommerce/woocommerce/pull/61325/commits/087287b4a52545afe175dda2c6f32faa9bffdd9a

RadoslavGeorgiev replied:
How about now?

RadoslavGeorgiev committed:
Lint fixes

RadoslavGeorgiev committed:
Add a tear-down, removing hooks

ebinnion committed:
Merge remote-tracking branch 'origin/trunk' into update/agentic-commerce-protocol-webhooks

ebinnion committed:
Update callback to only require one arg

ebinnion commented on plugins/woocommerce/src/Internal/Admin/Agentic/AgenticWebhookManager.php:
```diff
@@ -0,0 +1,238 @@
+			return $http_args;
+		}
+
+		// Compute HMAC signature per ACP webhook spec using WooCommerce's built-in method.
+		// The signature must be computed over the raw request body.
+		if ( isset( $http_args['body'] ) && ! empty( $webhook->get_secret() ) ) {
+			// Use WooCommerce's signature generation to ensure consistency.
+			$signature = $webhook->generate_signature( $http_args['body'] );
+
+			// Add Merchant-Signature header per ACP webhook specification.
+			$http_args['headers']['Merchant-Signature'] = $signature;
+		}
```
I'm not sure that the fact a filter could hook in and change this is good enough reason to not use the core function.

RadoslavGeorgiev reviewed (approved):
As mentioned earlier, there's a lot of heavy weight being lifted in this PR! 💪 

The code looks good and functions as expected. There were a couple of spin-off topics (when to send webhooks and how to set them up) for which I've created separate issues.
